banyan/src/forest/read.rs

```diff
@@ -19,7 +13,260 @@ use anyhow::{anyhow, Result};
+            };
+
+            //  Branch is exhausted: Ascend.
+            if head.is_exhausted(&self.mode) {
+                // Ascend to parent's node
+                self.stack.pop();
+
+                // increase last stack ptr, if there is still something left to
+                // traverse
+                if !self.stack.is_empty() {
+                    let last = self.stack.last_mut().expect("not empty");
+                    last.next_pos(&self.mode);
+                }
+                continue;
+            }
+
+            match self.forest.load_node(&head.index) {
+                Ok(NodeInfo::Branch(index, branch)) => {
+                    // we hit this branch node for the first time. Apply the
+                    // query on its children and store it
+                    if head.filter.is_empty() {
+                        head.filter = smallvec![true; index.summaries.len()];
+                        head.position = match self.mode {
+                            Mode::Forward => 0,
+                            Mode::Backward => branch.children.len() as isize - 1,
+                        };
+                        let start_offset = match self.mode {
+                            Mode::Forward => self.offset,
+                            Mode::Backward => self.offset - index.count,
+                        };
+                        self.query
+                            .intersecting(start_offset, index, &mut head.filter);
+                        debug_assert_eq!(branch.children.len(), head.filter.len());
+                    }
+
+                    let next_idx = head.position as usize;
+                    if head.filter[next_idx] {
+                        // Descend into next child
+                        self.stack.push(TraverseState::new(
+                            // TODO: clone :-( ?
+                            Arc::new(branch.children[next_idx].clone()),
+                        ));
+                        continue;
+                    } else {
+                        let index = &branch.children[next_idx];
+                        let range = match self.mode {
+                            Mode::Forward => {
+                                self.offset += index.count();
+                                self.offset - index.count()..self.offset
+                            }
+                            Mode::Backward => {
+                                self.offset -= index.count();
+                                self.offset..self.offset + index.count()
+                            }
+                        };
+                        head.next_pos(&self.mode);
+                        let placeholder: FilteredChunk<T, V, E> = FilteredChunk {
+                            range,
+                            data: Vec::new(),
+                            extra: (self.mk_extra)(index.as_index_ref()),
+                        };
+
+                        break placeholder;
```

Comment:
why is it necessary to emit these?

Reply:
the placeholders?

If you skip a subtree because either the filter does not match, or the subtree has been purged, you still want a placeholder to indicate that you have skipped offsets x..x+n. There is also the mk_extra fn that can extract some info from the tree node, e.g. the last lamport.

Reply:
Note that iter_filtered_chunked is a low level API that can be used in multiple ways. If you want just the data, you can just flatten them away.